### PR TITLE
[SYCL] Use unnamed lambdas to name fill/copy/etc. kernels when possible

### PR DESCRIPTION
**Pull request by aelovikov-intel**
This is an attempt to fix https://github.com/intel/llvm/issues/469 in a scenario when unnamed lambda feature is enabled.

### CODE REVIEW

File: sycl/test-e2e/Basic/memop/memop_unnamed_lambda.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+  // Same layout but different name to ensure no kernel name collisions happen.
+  struct Simple2 {
+    int c;
+  };
+
+  sycl::queue q;
+  void *p = sycl::aligned_alloc_device(alignof(Simple), sizeof(Simple), q);
+
+  q.fill(p, Simple{1}, 1).wait();
```
Comment (KseniyaTikhomirova) [inline sycl/test-e2e/Basic/memop/memop_unnamed_lambda.cpp:25]: @aelovikov-intel will it work if struct Simple and Simple2 have different sizes? Do we see two different kernels in device code. Could we add a test with nested type as in original reproducer or it is not expected to work?

Comment (aelovikov-intel) [reply sycl/test-e2e/Basic/memop/memop_unnamed_lambda.cpp:25]:
```
$ onetrace -d ./a.out 2>&1 | c++filt
...
typeinfo name for sycl::_V1::handler::fill<main::Simple>(void*, main::Simple const&, unsigned long)::{lambda(sycl::_V1::id<1>)#1}, ...
typeinfo name for sycl::_V1::handler::fill<main::Simple2>(void*, main::Simple2 const&, unsigned long)::{lambda(sycl::_V1::id<1>)#1}, ...
```

As such, it doesn't matter what exactly the struct is. I only added the same-size structs to highlight that we can't use `sizeof(Type)` as was supposed elsewhere as that will fail due to kernel name conflicts.

Comment (KseniyaTikhomirova) [reply sycl/test-e2e/Basic/memop/memop_unnamed_lambda.cpp:25]: ok, got it. what is about nested type testing request?

**Review by KseniyaTikhomirova — APPROVED** [review]
LGTM

**Pull request merged**